README.md
PdbParser
==========

Experimental library for parsing PDB files.

<br><b>status:</b> in development.  Not yet posted for sharing

License
---------
Apache.  <i>See LICENSE.txt</i>

Credits
-------
The focus of the PDB work is on study of PDB files hosted by the
[RCSB][1].  Testing of the work has been carried out on a subset of
PDB files featured in the 20 years of the <b>PDB101</b> monthly 
online publication also hosted by the [RCSB][1].

  [1]: https://www.rcsb.org/

- Bond information

The bond table information derives from pdbe database information, 
for example from the query:

    https://www.ebi.ac.uk/pdbe/api/pdb/compound/bonds/dg
    
For more information, see:

https://www.ebi.ac.uk/about/terms-of-use/

The bond table has been debugged using the PDB101 sample subset to include 
bonds not present in the EBI dataset.
    
Build technology credits
------------------------

- jakewharton/Picnic - for nice maven central publication structure

- MarcinMoskala/KotlinDiscreteMathToolkit - for nice testing structure

Documentation technology credits
--------------------------------

The documentation uses `MkDocs` framework and the `material` theme provided 
kindly by Martin Donath (github/squidfunk).

https://squidfunk.github.io/mkdocs-material/
<br>https://www.mkdocs.org/

    
 


CHANGELOG.md
Change Log
==========

Version 1.0.4 
----------------------------
 * Change from Double to Float atom coordinates

Version 1.0.3
----------------------------

* First SNAPSHOT version

Version 1.0.0 *(2020-06-26)*
----------------------------

 * Initial version


pdbparser/src/test/java/com/kotmol/pdbParser/TERtest01.kt
/*
 *  Copyright 2020-2023 James Andreas
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *       http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License
 */

@file:Suppress(
    "unused",
    "unused_variable",
    "unused_parameter",
    "unused_property",
    "deprecation",
    "ConstantConditionIf",
    "LocalVariableName",
    "PropertyName")

package com.kotmol.pdbParser

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import com.kotmol.pdbParser.PdbAtom.AtomType.IS_TER_RECORD

// https://blog.jetbrains.com/idea/2016/08/using-junit-5-in-intellij-idea/

internal class TERtest01 {

    /**
     * 1ZNF - a Zinc Finger PDB with 37 models (edited)
     *    What is unique about this test case is that the same
     *    residue number is used across a TER record.  The residue names
     *    change.   The TER record asserts that the chain ends there.
     */
    lateinit var str : ByteArrayInputStream
    private lateinit var aModelTest : String
    @org.junit.jupiter.api.BeforeEach
    fun setUp() {
        val theModelString = """
         1         2         3         4         5         6         7
12345678901234567890123456789012345678901234567890123456789012345678901234567890"""
        aModelTest = """
ATOM    204  P    DT E  12      41.677 -64.307   5.732  1.00122.09           P  
ATOM    205  OP1  DT E  12      41.863 -63.536   6.997  1.00122.09           O  
ATOM    206  OP2  DT E  12      40.540 -65.263   5.602  1.00122.09           O  
ATOM    207  O5'  DT E  12      41.490 -63.265   4.544  1.00122.09           O  
ATOM    208  C5'  DT E  12      41.284 -63.725   3.216  1.00122.09           C  
ATOM    209  C4'  DT E  12      41.127 -62.551   2.284  1.00122.09           C  
ATOM    210  O4'  DT E  12      42.352 -61.765   2.361  1.00122.09           O  
ATOM    211  C3'  DT E  12      41.055 -63.092   0.847  1.00122.09           C  
ATOM    212  O3'  DT E  12      40.589 -62.225  -0.194  1.00122.09           O  
ATOM    213  C2'  DT E  12      42.507 -63.144   0.452  1.00122.09           C  
ATOM    214  C1'  DT E  12      42.986 -61.878   1.111  1.00122.09           C  
ATOM    215  N1   DT E  12      44.435 -61.796   1.185  1.00122.09           N  
ATOM    216  C2   DT E  12      45.194 -62.908   1.513  1.00122.09           C  
ATOM    217  O2   DT E  12      44.746 -63.884   2.121  1.00121.05           O  
ATOM    218  N3   DT E  12      46.495 -62.834   1.058  1.00122.05           N  
ATOM    219  C4   DT E  12      47.050 -61.795   0.349  1.00122.09           C  
ATOM    220  O4   DT E  12      48.112 -61.938  -0.174  1.00122.09           O  
ATOM    221  C5   DT E  12      46.248 -60.610   0.228  1.00122.09           C  
ATOM    222  C7   DT E  12      46.823 -59.410  -0.452  1.00122.09           C  
ATOM    223  C6   DT E  12      45.009 -60.650   0.696  1.00122.09           C  
TER     224       DT E  12                                                      
ATOM    225  O5'  DA D  12      57.253 -61.992  -0.544  1.00117.97           O  
ATOM    226  C5'  DA D  12      57.447 -61.049  -1.619  1.00120.06           C  
ATOM    227  C4'  DA D  12      56.212 -60.189  -1.578  1.00122.09           C  
ATOM    228  O4'  DA D  12      55.088 -61.095  -1.431  1.00122.09           O  
ATOM    229  C3'  DA D  12      56.193 -59.244  -0.360  1.00122.09           C  
ATOM    230  O3'  DA D  12      55.679 -57.956  -0.705  1.00122.09           O  
ATOM    231  C2'  DA D  12      55.383 -59.973   0.710  1.00122.09           C  
ATOM    232  C1'  DA D  12      54.568 -60.901  -0.144  1.00122.09           C  
ATOM    233  N9   DA D  12      53.641 -61.943   0.293  1.00122.09           N  
ATOM    234  C8   DA D  12      53.592 -63.296   0.675  1.00122.09           C  
ATOM    235  N7   DA D  12      52.407 -63.697   1.027  1.00122.09           N  
ATOM    236  C5   DA D  12      51.648 -62.522   0.878  1.00122.09           C  
ATOM    237  C6   DA D  12      50.290 -62.250   1.103  1.00122.09           C  
ATOM    238  N6   DA D  12      49.433 -63.180   1.501  1.00122.09           N  
ATOM    239  N1   DA D  12      49.855 -60.973   0.861  1.00122.09           N  
ATOM    240  C2   DA D  12      50.735 -60.076   0.425  1.00122.09           C  
ATOM    241  N3   DA D  12      52.029 -60.220   0.194  1.00122.09           N  
ATOM    242  C4   DA D  12      52.392 -61.477   0.435  1.00122.09           C  
        """.trimIndent()

        str = aModelTest.byteInputStream()

    }

    @org.junit.jupiter.api.AfterEach
    fun tearDown() {
        str.close()
    }

    /**
     *
     */
    @Test
    @DisplayName( "test processing of a TER record")
    fun testModelSkipping() {

        val mol = Molecule()
        val messages : MutableList<String> = mutableListOf()

        ParserPdbFile
                .Builder(mol)
                .setMoleculeName("1ZNF")
                .setMessageStrings(messages)
                .loadPdbFromStream(str)
                .parse()

        // two residues separated by a TER record (see above)
        val atoms = mol.atomNumberToAtomInfoHash
        assertEquals(39, atoms.size)

        // bonds from the two records don't overlap
        assertEquals(41, mol.bondList.size)

        // check that the TER record is entered into the
        // atom list as a TER_RECORD type.
        val shouldBeTER = mol.atomNumberToAtomInfoHash[224]
        assertNotNull(shouldBeTER)
        assertEquals(IS_TER_RECORD, shouldBeTER!!.atomType)

    }
}

pdbparser/src/test/java/com/kotmol/pdbParser/AFailedTest.kt
/*
 *  Copyright 2020-2023 James Andreas
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *       http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License
 */

package com.kotmol.pdbParser

import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

internal class AFailedTest {
    @Test @Disabled
    @DisplayName("This Test Fails by design")
    fun failTest() {
        assert(false)
    }
}

pdbparser/src/main/java/com/kotmol/pdbParser/AtomInformationTable.kt
/*
 *  Copyright 2020-2023 James Andreas
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *       http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License
 */

@file:Suppress(
        "ArrayInDataClass",
        "MemberVisibilityCanBePrivate", "UnnecessaryVariable", "ReplaceJavaStaticMethodWithKotlinAnalog",
        "unused", "UNUSED_VARIABLE", "ReplaceManualRangeWithIndicesCalls", "UNUSED_VALUE", "ReplaceWithOperatorAssignment",
        "UNUSED_PARAMETER", "UNUSED_CHANGED_VALUE", "CanBeVal", "SimplifyBooleanWithConstants",
        "ConvertTwoComparisonsToRangeCheck", "ReplaceSizeCheckWithIsNotEmpty", "LiftReturnOrAssignment",
        "VARIABLE_WITH_REDUNDANT_INITIALIZER"
)
package com.kotmol.pdbParser

object AtomInformationTable {

    // generated from JimConvertAtomInfoTable.kt (IntelliJ Idea)
    data class AtomNameNumber(
            var number: Int = 0,
            var symbol: String = "",
            var name: String = "",
            var vanDerWaalsRadius: Int = 300, // default size if not known
            var color: FloatArray = floatArrayOf(0.0f, 0.0f, 0.0f, 1.0f)
    )

    val atomSymboltoAtomNumNameColor = HashMap<String, AtomNameNumber>()

    val atomTable = listOf(
            AtomNameNumber(number = 1, symbol = "H", name = "hydrogen", vanDerWaalsRadius = 120, color = floatArrayOf(1.0f, 1.0f, 1.0f, 1.0f)),
            AtomNameNumber(number = 2, symbol = "HE", name = "helium", vanDerWaalsRadius = 140, color = floatArrayOf(0.8509804f, 1.0f, 1.0f, 1.0f)),
            AtomNameNumber(number = 3, symbol = "LI", name = "lithium", vanDerWaalsRadius = 182, color = floatArrayOf(0.69803923f, 0.12941177f, 0.12941177f, 1.0f)),
            AtomNameNumber(number = 4, symbol = "BE", name = "beryllium", vanDerWaalsRadius = 153, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 5, symbol = "B", name = "boron", vanDerWaalsRadius = 192, color = floatArrayOf(0.0f, 1.0f, 0.0f, 1.0f)),
            AtomNameNumber(number = 6, symbol = "C", name = "carbon", vanDerWaalsRadius = 170, color = floatArrayOf(0.8901961f, 0.7647059f, 0.59607846f, 1.0f)),
            AtomNameNumber(number = 7, symbol = "N", name = "nitrogen", vanDerWaalsRadius = 155, color = floatArrayOf(0.5294118f, 0.80784315f, 0.9019608f, 1.0f)),
            AtomNameNumber(number = 8, symbol = "O", name = "oxygen", vanDerWaalsRadius = 152, color = floatArrayOf(1.0f, 0.0f, 0.0f, 1.0f)),
            AtomNameNumber(number = 9, symbol = "F", name = "fluorine", vanDerWaalsRadius = 147, color = floatArrayOf(0.85490197f, 0.64705884f, 0.1254902f, 1.0f)),
            AtomNameNumber(number = 10, symbol = "NE", name = "neon", vanDerWaalsRadius = 154, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 11, symbol = "NA", name = "sodium", vanDerWaalsRadius = 227, color = floatArrayOf(0.0f, 0.0f, 1.0f, 1.0f)),
            AtomNameNumber(number = 12, symbol = "MG", name = "magnesium", vanDerWaalsRadius = 173, color = floatArrayOf(0.13333334f, 0.54509807f, 0.13333334f, 1.0f)),
            AtomNameNumber(number = 13, symbol = "AL", name = "aluminium", vanDerWaalsRadius = 184, color = floatArrayOf(0.4117647f, 0.4117647f, 0.4117647f, 1.0f)),
            AtomNameNumber(number = 14, symbol = "SI", name = "silicon", vanDerWaalsRadius = 210, color = floatArrayOf(0.85490197f, 0.64705884f, 0.1254902f, 1.0f)),
            AtomNameNumber(number = 15, symbol = "P", name = "phosphorus", vanDerWaalsRadius = 180, color = floatArrayOf(1.0f, 0.6666667f, 0.0f, 1.0f)),
            AtomNameNumber(number = 16, symbol = "S", name = "sulfur", vanDerWaalsRadius = 180, color = floatArrayOf(1.0f, 1.0f, 0.0f, 1.0f)),
            AtomNameNumber(number = 17, symbol = "CL", name = "chlorine", vanDerWaalsRadius = 175, color = floatArrayOf(0.0f, 1.0f, 0.0f, 1.0f)),
            AtomNameNumber(number = 18, symbol = "AR", name = "argon", vanDerWaalsRadius = 188, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 19, symbol = "K", name = "potassium", vanDerWaalsRadius = 275, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 20, symbol = "CA", name = "calcium", vanDerWaalsRadius = 231, color = floatArrayOf(0.4117647f, 0.4117647f, 0.4117647f, 1.0f)),
            AtomNameNumber(number = 21, symbol = "SC", name = "scandium", vanDerWaalsRadius = 211, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 22, symbol = "TI", name = "titanium", vanDerWaalsRadius = 300, color = floatArrayOf(0.4117647f, 0.4117647f, 0.4117647f, 1.0f)),
            AtomNameNumber(number = 23, symbol = "V", name = "vanadium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 24, symbol = "CR", name = "chromium", vanDerWaalsRadius = 300, color = floatArrayOf(0.4117647f, 0.4117647f, 0.4117647f, 1.0f)),
            AtomNameNumber(number = 25, symbol = "MN", name = "manganese", vanDerWaalsRadius = 300, color = floatArrayOf(0.4117647f, 0.4117647f, 0.4117647f, 1.0f)),
            AtomNameNumber(number = 26, symbol = "FE", name = "iron", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.6666667f, 0.0f, 1.0f)),
            AtomNameNumber(number = 27, symbol = "CO", name = "cobalt", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 28, symbol = "NI", name = "nickel", vanDerWaalsRadius = 163, color = floatArrayOf(0.5019608f, 0.15686275f, 0.15686275f, 1.0f)),
            AtomNameNumber(number = 29, symbol = "CU", name = "copper", vanDerWaalsRadius = 140, color = floatArrayOf(0.5019608f, 0.15686275f, 0.15686275f, 1.0f)),
            AtomNameNumber(number = 30, symbol = "ZN", name = "zinc", vanDerWaalsRadius = 139, color = floatArrayOf(0.5019608f, 0.15686275f, 0.15686275f, 1.0f)),
            AtomNameNumber(number = 31, symbol = "GA", name = "gallium", vanDerWaalsRadius = 187, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 32, symbol = "GE", name = "germanium", vanDerWaalsRadius = 211, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 33, symbol = "AS", name = "arsenic", vanDerWaalsRadius = 185, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 34, symbol = "SE", name = "selenium", vanDerWaalsRadius = 190, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 35, symbol = "BR", name = "bromine", vanDerWaalsRadius = 185, color = floatArrayOf(0.5019608f, 0.15686275f, 0.15686275f, 1.0f)),
            AtomNameNumber(number = 36, symbol = "KR", name = "krypton", vanDerWaalsRadius = 202, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 37, symbol = "RB", name = "rubidium", vanDerWaalsRadius = 303, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 38, symbol = "SR", name = "strontium", vanDerWaalsRadius = 249, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 39, symbol = "Y", name = "yttrium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 40, symbol = "ZR", name = "zirconium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 41, symbol = "NB", name = "niobium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 42, symbol = "MO", name = "molybdenum", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 43, symbol = "TC", name = "technetium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 44, symbol = "RU", name = "ruthenium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 45, symbol = "RH", name = "rhodium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 46, symbol = "PD", name = "palladium", vanDerWaalsRadius = 163, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 47, symbol = "AG", name = "silver", vanDerWaalsRadius = 172, color = floatArrayOf(0.4117647f, 0.4117647f, 0.4117647f, 1.0f)),
            AtomNameNumber(number = 48, symbol = "CD", name = "cadmium", vanDerWaalsRadius = 158, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 49, symbol = "IN", name = "indium", vanDerWaalsRadius = 193, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 50, symbol = "SN", name = "tin", vanDerWaalsRadius = 217, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 51, symbol = "SB", name = "antimony", vanDerWaalsRadius = 206, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 52, symbol = "TE", name = "tellurium", vanDerWaalsRadius = 206, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 53, symbol = "I", name = "iodine", vanDerWaalsRadius = 198, color = floatArrayOf(0.627451f, 0.1254902f, 0.9411765f, 1.0f)),
            AtomNameNumber(number = 54, symbol = "XE", name = "xenon", vanDerWaalsRadius = 216, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 55, symbol = "CS", name = "caesium", vanDerWaalsRadius = 343, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 56, symbol = "BA", name = "barium", vanDerWaalsRadius = 268, color = floatArrayOf(1.0f, 0.6666667f, 0.0f, 1.0f)),
            AtomNameNumber(number = 57, symbol = "LA", name = "lanthanum", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 58, symbol = "CE", name = "cerium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 59, symbol = "PR", name = "praseodymium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 60, symbol = "ND", name = "neodymium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 61, symbol = "PM", name = "promethium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 62, symbol = "SM", name = "samarium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 63, symbol = "EU", name = "europium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 64, symbol = "GD", name = "gadolinium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 65, symbol = "TB", name = "terbium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 66, symbol = "DY", name = "dysprosium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 67, symbol = "HO", name = "holmium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 68, symbol = "ER", name = "erbium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 69, symbol = "TM", name = "thulium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 70, symbol = "YB", name = "ytterbium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 71, symbol = "LU", name = "lutetium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 72, symbol = "HF", name = "hafnium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 73, symbol = "TA", name = "tantalum", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 74, symbol = "W", name = "tungsten", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 75, symbol = "RE", name = "rhenium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 76, symbol = "OS", name = "osmium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 77, symbol = "IR", name = "iridium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 78, symbol = "PT", name = "platinum", vanDerWaalsRadius = 175, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 79, symbol = "AU", name = "gold", vanDerWaalsRadius = 166, color = floatArrayOf(0.85490197f, 0.64705884f, 0.1254902f, 1.0f)),
            AtomNameNumber(number = 80, symbol = "HG", name = "mercury", vanDerWaalsRadius = 155, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 81, symbol = "TL", name = "thallium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 82, symbol = "PB", name = "lead", vanDerWaalsRadius = 202, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 83, symbol = "BI", name = "bismuth", vanDerWaalsRadius = 207, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 84, symbol = "PO", name = "polonium", vanDerWaalsRadius = 197, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 85, symbol = "AT", name = "astatine", vanDerWaalsRadius = 202, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 86, symbol = "RN", name = "radon", vanDerWaalsRadius = 220, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 87, symbol = "FR", name = "francium", vanDerWaalsRadius = 348, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 88, symbol = "RA", name = "radium", vanDerWaalsRadius = 283, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 89, symbol = "AC", name = "actinium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 90, symbol = "TH", name = "thorium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 91, symbol = "PA", name = "protactinium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 92, symbol = "U", name = "uranium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 93, symbol = "NP", name = "neptunium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 94, symbol = "PU", name = "plutonium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 95, symbol = "AM", name = "americium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 96, symbol = "CM", name = "curium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 97, symbol = "BK", name = "berkelium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 98, symbol = "CF", name = "californium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 99, symbol = "ES", name = "einsteinium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 100, symbol = "FM", name = "fermium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 101, symbol = "MD", name = "mendelevium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 102, symbol = "NO", name = "nobelium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 103, symbol = "LR", name = "lawrencium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 104, symbol = "RF", name = "rutherfordium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 105, symbol = "DB", name = "dubnium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 106, symbol = "SG", name = "seaborgium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 107, symbol = "BH", name = "bohrium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 108, symbol = "HS", name = "hassium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 109, symbol = "MT", name = "meitnerium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 110, symbol = "DS", name = "darmstadtium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 111, symbol = "RG", name = "roentgenium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 112, symbol = "CN", name = "copernicium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 113, symbol = "UUT", name = "ununtrium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 114, symbol = "FL", name = "flerovium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 115, symbol = "UUP", name = "ununpentium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 116, symbol = "LV", name = "livermorium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 117, symbol = "UUS", name = "ununseptium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f)),
            AtomNameNumber(number = 118, symbol = "UUO", name = "ununoctium", vanDerWaalsRadius = 300, color = floatArrayOf(1.0f, 0.078431375f, 0.5764706f, 1.0f))
    )

    init {
        val iter = atomTable.iterator()
        while (iter.hasNext()) {
            val thisAtom = iter.next()
            atomSymboltoAtomNumNameColor[thisAtom.symbol] = thisAtom
        }
    }
}
